src/integrations/gl/sdl_gl_platform.cpp

```diff
@@ -1,19 +1,168 @@
 #include "platform/gpu.h"
 #include "common/error.h"
+#include "common/macros.h"
 
 #include "SDL.h"
 #include "GL/gl3w.h"
 
+#include <cstdlib>
+
+namespace vx
+{
 namespace
 {
+struct gl_buffer
+{
+    GLuint object;
+    GLenum target;
+};
+
+static_assert(sizeof(gl_buffer) == sizeof(uptr), "gl_buffer is not the size of a pointer");
+
+struct gl_texture
+{
+    GLuint object;
+    GLenum target;
+};
+
+static_assert(sizeof(gl_texture) == sizeof(uptr), "gl_texture is not the size of a pointer");
+
+struct gl_shader
+{
+    GLuint object;
+    GLuint padding;
+};
+
+static_assert(sizeof(gl_shader) == sizeof(uptr), "gl_shader is not the size of a pointer");
+
+struct gl_pipeline
+{
+    GLuint program;
+
+    bool depth_test_enabled;
+    bool culling_enabled;
+    bool scissor_test_enabled;
+    bool blend_enabled;
+
+    // TODO: blend function, comparison function?
+};
+
+static_assert(sizeof(gl_pipeline) == sizeof(uptr), "gl_pipeline is not the size of a pointer");
+
+struct gl_attribute
+{
+    GLint index;
+    GLint elements;
+    GLenum type;
+    GLboolean normalized;
+    const void* offset;
+};
+
+struct gl_descriptor
+{
+    gl_attribute* attributes;
+    u32 attribute_count;
+    u32 stride;
+};
+
 struct gl_device
 {
-    SDL_GLContext ctx;
+    SDL_GLContext context;
+    GLuint dummy_vao;
+
+    int2 display_size;
+    float2 display_scale;
+
+    gl_pipeline current_pipeline;
 };
+
+GLint gpu_convert_enum(gpu_buffer_type type)
+{
+    switch (type)
+    {
+        case gpu_buffer_type::vertex:
+            return GL_ARRAY_BUFFER;
+        case gpu_buffer_type::index:
+            return GL_ELEMENT_ARRAY_BUFFER;
+        case gpu_buffer_type::shader_storage:
+            return GL_SHADER_STORAGE_BUFFER;
+        case gpu_buffer_type::constant:
+            return GL_UNIFORM_BUFFER;
+        default:
+            assert(!"Invalid gpu_buffer_type");
```

Comment:
Convert these to fatal() and output the `type` while you're at it. Better diagnostics.